### PR TITLE
Addresses Issue #615 : Adding new parameters "Colorbar" and "Heatmap" in PCA visualizer

### PR DESCRIPTION
**Pull request by percygautam**
The following changes were made to `pca.py` referenced in #615 :

- Added new optional parameters 'Colorbar' and 'Heatmap' which shows the magnitude of each feature value to the component, for both '2D' and '3D' cases.

Next Step:

- Updating Documentation with a proper explanation on how to use these two optional parameters when using PCA visualizer.

P.S: I am adding a snippet showing PCA visualization of the dataset credit from yellowbrick.datasets, when both the parameters are True and proj_dim = 2.

sd

### CODE REVIEW

**Comment by percygautam** [comment]
@rebeccabilbro Hey, there. 
I have corrected the `transform` method, which now works with ax object passed. I am getting difficulties in getting the axis for heatmap to set x and y labels in the `finalize` method. Currently, I am using `plt.gca()` method to get the axis. How can I fix this issue?

**Comment by naresh-bachwani** [comment]
Hey @percygautam, 
I will be working on PCA this summer, so I thought to have a look at your solution. You have done a great job! To answer your query can you try using something like `self.ax1.set_xticks()`?( where ax1 is the axis of the heatmap) Let me know if it works. However, wait till @rebeccabilbro reviews it. She may have something to add.

**Comment by rebeccabilbro** [comment]
Hello there @percygautam and so glad to see this moving forward! 

My recommendation is definitely along the lines of what @naresh-bachwani is suggesting (P.S. thanks for jumping in here @naresh-bachwani and very glad you're keeping track of this so that you'll be ready to jump in with the other PCA roadmap improvements once this is merged! Potentially you and I should team up in reviewing this PR once @percygautam is ready for us!). 

@percygautam I propose that you read through our latest improvements to `JointPlotVisualizer`, take a look at what we do to divide up the axes (depending on which params the user passes in), and come up with a similar solution. In `JointPlot` we have a [two axes](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/features/jointplot.py#L142) that are `None` by default (though probably for PCA you'd call them something like `self.uax` and `self.lax` for "upper axes" and "lower axes"). We also have a [private method](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/features/jointplot.py#L208) that divides the axes up [iff certain parameters have been set by the user](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/features/jointplot.py#L178-L179) (e.g. `colorbar` and `heatmap` in the case of PCA). You can call this method on `init`, after which you will be able to reference `self.uax` and `self.lax` in `finalize` to add whatever finishing touches are needed (e.g. `self.uax.set_xticks()` like @naresh-bachwani suggested).

Hope this helps and keep us posted!

**Comment by percygautam** [comment]
Hey @rebeccabilbro @naresh-bachwani  
Thanks, for your suggestions. I have used the methods used for JointPlot and now the heatmap and colorbar are rendering properly for 2d projections. But this method will not work for 3d projections as we can't get the axis of 3d scatter plot. Currently, I am using subplot for rendering 3d plot, but this method is not feasible! Any suggestions for 3d projections?

**Comment by rebeccabilbro** [comment]
> Thanks, for your suggestions. I have used the methods used for JointPlot and now the heatmap and colorbar are rendering properly for 2d projections. But this method will not work for 3d projections as we can't get the axis of 3d scatter plot. Currently, I am using subplot for rendering 3d plot, but this method is not feasible! Any suggestions for 3d projections?

I propose adding logic so that the multi-axes are only available if the user selects a 2D plot (e.g. raise an error if `proj_dim == 3`, and either `heatmap == True` or `colorbar == True`). 

Also, @percygautam &mdash; it looks like for some reason your PR has integrated commits from an unrelated PR for the new Cook's Distance visualizer. Would you please remove all of [those additions](https://github.com/DistrictDataLabs/yellowbrick/pull/884/commits/9c1cf534ea8f82a686a1fe786a0fed0b86e95bb7) from this PR so that we will not have a merge conflict?

**Comment by percygautam** [comment]
@rebeccabilbro  I have done the changes you mentioned.

**Comment by rebeccabilbro** [comment]
Ok, great @percygautam! I've just pulled from `develop` to make sure your branch is synced up; once the automated tests have finished running I'll start my review!

File: yellowbrick/features/pca.py [hunk 1]
```diff
@@ -152,15 +188,21 @@ def fit(self, X, y=None, **kwargs):
         return self
 
     def transform(self, X, y=None, **kwargs):
+        self.orig_X = X
```
[the remark this answers is not in the document]
Comment (percygautam) [reply yellowbrick/features/pca.py]: Yes, you are right. I had added it to get feature names.

**Comment by percygautam** [comment]
@naresh-bachwani  Thank you for your suggestions. I have done the changes you mentioned.

**Comment by naresh-bachwani** [comment]
Hey there @rebeccabilbro and @percygautam,
Thank you @percygautam  for making the changes. The tests are looking good. Just a few minor changes and we will have the tests passing. As @rebeccabilbro said, these tests are failing (probably) due to text differences. So, we just need to remove the text. To achieve that you can remove the finalize calls from the test. Or you can also explicitly set ticks to `None`.

@rebeccabilbro, we added the following tests:
1. Test heatmap
2. Test colorbar
3. Test heatmap and colorbar
4. Test the error when the projection dimension is 3 and heatmap or colorbar is true.

Are there any other tests that you think we must be adding? Apart from this and docstring, all other things are looking great!

File: yellowbrick/features/pca.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (naresh-bachwani) [reply yellowbrick/features/pca.py]: Hmm great! We could have used `self.features_` instead. The super class `MultiFeatureVisualizer` provides this functionality.

**Comment by bbengfort** [comment]
@rebeccabilbro @naresh-bachwani -- it looks like the text removal tools in the image similarity tests are only affecting the primary Axes - and the text that is causing the issues here are on the two new axes of the colorbar and the heatmap. I've opened up an issue #916 that will address this. 

@percygautam - For now, would you mind removing the ticks explicitly in the tests as @naresh-bachwani suggested? Please put a comment with a `# TODO: explicity removal should be removed after #916 is fixed` in the places that you make that modification.

**Comment by rebeccabilbro** [comment]
@percygautam &mdash; just checking in, how's your progress on this? We're so close!

**Comment by percygautam** [comment]
@rebeccabilbro Sorry for the delay in progress! Thank you, @naresh-bachwani  and @bbengfort for your inputs. I'll do the changes mentioned by today!

**Comment by percygautam** [comment]
@rebeccabilbro I have done some changes! I am having some problem with pytest, will change the baseline images once it's resolved.

**Comment by percygautam** [comment]
@rebeccabilbro  @naresh-bachwani  After removing text from the images, tests are still failing. How to rectify this?

**Pull request merged**